### PR TITLE
Convert "blocks" table to utf8mb4 to accept 4 byte utf8 characters

### PR DESCRIPTION
**Pull request by joshlory**
Also changed the "category" column from `t.text` -> `t.string`. Category should be a short string.

After migration:

```
mysql> show full columns from blocks;
+-------------+--------------+--------------------+------+-----+---------+----------------+---------------------------------+---------+
| Field       | Type         | Collation          | Null | Key | Default | Extra          | Privileges                      | Comment |
+-------------+--------------+--------------------+------+-----+---------+----------------+---------------------------------+---------+
| id          | int(11)      | NULL               | NO   | PRI | NULL    | auto_increment | select,insert,update,references |         |
| name        | varchar(255) | utf8mb4_unicode_ci | YES  |     | NULL    |                | select,insert,update,references |         |
| level_type  | varchar(255) | utf8mb4_unicode_ci | YES  |     | NULL    |                | select,insert,update,references |         |
| category    | varchar(255) | utf8mb4_unicode_ci | YES  |     | NULL    |                | select,insert,update,references |         |
| config      | text         | utf8mb4_unicode_ci | YES  |     | NULL    |                | select,insert,update,references |         |
| helper_code | text         | utf8mb4_unicode_ci | YES  |     | NULL    |                | select,insert,update,references |         |
| created_at  | datetime     | NULL               | NO   |     | NULL    |                | select,insert,update,references |         |
| updated_at  | datetime     | NULL               | NO   |     | NULL    |                | select,insert,update,references |         |
+-------------+--------------+--------------------+------+-----+---------+----------------+---------------------------------+---------+
8 rows in set (0.00 sec)
```

### CODE REVIEW

File: dashboard/db/migrate/20180703165502_change_blocks_to_utf8mb4.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    execute 'alter table blocks modify config text(16383) charset utf8mb4 collate utf8mb4_unicode_ci'
+    execute 'alter table blocks modify helper_code text(16383) charset utf8mb4 collate utf8mb4_unicode_ci'
+  end
+
+  def down
+    execute 'alter table blocks convert to charset utf8 collate utf8_unicode_ci'
+    execute 'alter table blocks modify name varchar(255) charset utf8 collate utf8_unicode_ci'
+    execute 'alter table blocks modify category varchar(255) charset utf8 collate utf8_unicode_ci'
+    execute 'alter table blocks modify config text(21845) charset utf8 collate utf8_unicode_ci'
+    execute 'alter table blocks modify helper_code text(21845) charset utf8 collate utf8_unicode_ci'
```
Comment (joshlory) [inline dashboard/db/migrate/20180703165502_change_blocks_to_utf8mb4.rb:15]: Rails annotates the model as `text(65355)` but MySQL wants the value to be the actual number of characters stored.  65355 / 3 = 21845.

**Comment by joshlory** [comment]
[Trying to add emoji to a block definition shouldn't break the deploy pipeline](https://trello.com/c/LylfOzeX/189-trying-to-add-emoji-to-a-block-definition-shouldnt-break-the-deploy-pipeline)

File: dashboard/app/models/block.rb [hunk 2]
```diff
@@ -5,7 +5,7 @@
 #  id          :integer          not null, primary key
 #  name        :string(255)
 #  level_type  :string(255)
-#  category    :text(65535)
+#  category    :string(255)
```
Comment (joshlory) [inline dashboard/app/models/block.rb]: Intentional change.  This should be a string field.

Comment (joshlory) [inline dashboard/app/models/block.rb:8]: Intentional change.  Category should be a short string.

**Comment by wjordan** [comment]
Migration looks good. I thought handling utf8mb4 queries would also require changes to our mysql2 adapter's client-connection charset/encoding settings as well. This migration should be safe either way, but might not be effective without further changes to the client. In any case, it would be good to test/verify that this migration actually fixes emoji input in block definitions as expected.

**Comment by joshlory** [comment]
@wjordan good call, I updated the default adaptor to also support utf8mb4 in commit 2ba71c4647a1a1f8d2c65272a99c6d8b5f651884.  Are there any downsides to this change?

When testing locally, tables that have `utf8` encoding instead of `utf8mb4` still get the `Mysql2::Error: Incorrect string value` error when trying to save 4-byte emoji, but saving these the "blocks" table now works.

image

**Pull request merged**